cpp/src/transform/row_bit_count.cu

```diff
@@ -0,0 +1,508 @@
+  void operator()(column_view const& col,
+                  std::vector<cudf::column_view>& out,
+                  std::vector<column_info>& info,
+                  hierarchy_info& h_info,
+                  rmm::cuda_stream_view stream,
+                  size_type cur_depth,
+                  size_type cur_branch_depth,
+                  int parent_index)
+  {
+    out.push_back(col);
+    info.push_back({cur_depth, cur_branch_depth, cur_branch_depth});
+    h_info.complex_type_count++;
+  }
+
+  // lists
+  template <typename T, std::enable_if_t<std::is_same<T, list_view>::value>* = nullptr>
+  void operator()(column_view const& col,
+                  std::vector<cudf::column_view>& out,
+                  std::vector<column_info>& info,
+                  hierarchy_info& h_info,
+                  rmm::cuda_stream_view stream,
+                  size_type cur_depth,
+                  size_type cur_branch_depth,
+                  int parent_index)
+  {
+    // track branch depth as we reach this list and after we pass it
+    size_type const branch_depth_start = cur_branch_depth;
+    if (parent_index >= 0 && out[parent_index].type().id() == type_id::STRUCT) {
+      cur_branch_depth++;
+      if (cur_branch_depth > h_info.max_branch_depth) {
+        h_info.max_branch_depth = cur_branch_depth;
+      }
+    }
+    size_type const branch_depth_end = cur_branch_depth;
+
+    out.push_back(col);
+    info.push_back({cur_depth, branch_depth_start, branch_depth_end});
+
+    lists_column_view lcv(col);
+    auto iter = cudf::detail::make_counting_transform_iterator(
+      0, [col = lcv.get_sliced_child(stream)](auto i) { return col; });
+    h_info.complex_type_count++;
+
+    flatten_hierarchy(
+      iter, iter + 1, out, info, h_info, stream, cur_depth + 1, cur_branch_depth, out.size() - 1);
+  }
+
+  // structs
+  template <typename T, std::enable_if_t<std::is_same<T, struct_view>::value>* = nullptr>
+  void operator()(column_view const& col,
+                  std::vector<cudf::column_view>& out,
+                  std::vector<column_info>& info,
+                  hierarchy_info& h_info,
+                  rmm::cuda_stream_view stream,
+                  size_type cur_depth,
+                  size_type cur_branch_depth,
+                  int parent_index)
+  {
+    out.push_back(col);
+    info.push_back({cur_depth, cur_branch_depth, cur_branch_depth});
+
+    h_info.simple_per_row_size += col.nullable() ? 1 : 0;
+
+    structs_column_view scv(col);
+    auto iter = cudf::detail::make_counting_transform_iterator(
+      0, [&scv](auto i) { return scv.get_sliced_child(i); });
+    flatten_hierarchy(iter,
+                      iter + scv.num_children(),
+                      out,
+                      info,
+                      h_info,
+                      stream,
+                      cur_depth + 1,
+                      cur_branch_depth,
+                      out.size() - 1);
+  }
+
+  // everything else
```

Comment:
The question is:  what would the size of a "row" be for a dictionary?  You could gather 1 row and get 16 bytes of chars and 1  index.  But then gather 1000 rows, all of the same index,  but still only get 16 bytes of chars.  Maybe just dictionary should just be row size == sizeof(index) * num_rows and disregard chars entirely?  It puts a little more burden on the user, but that's already the case somewhat with lists/strings (because of the trailing offset)

Reply:
You could do a count on the indices and a count on the keys and add them up.
Note that the keys of a dictionary do not have to be strings.

Reply:
Oh, you said size of a _row_. Yes, I see that would be difficult.

Reply:
I came here with just this question. It should be possible to consider the size of a dictionary column as the size of the un-dictionaried data. It will give you a very pessimistic size of the data which should be sufficient for getting upper bound on the size of the table slices to join on.

This may or may not suit the libcudf ideology but if it does, I have a nice use case for this in cuIO.

Reply:
That may end up being ok.  This code ends up pessimistic in another case: string or list columns inside structs that have had their validity vectors modified after creation.   So row 4 for a string column might actually have 16 real chars of data.  But since the struct might have caused the null mask for that row to be set to 0, if you gathered that row, you would get 0 chars. 

I think the most important thing is that this never becomes an underestimate.

Reply:
Is there a resolution here?

Reply:
@devavret

Reply:
I'm not requesting changes here. This was just a suggestion. You can ignore if supporting dictionary columns is not a priority. ￼